### PR TITLE
College board data

### PR DESCRIPTION
**Pull request by bencodeorg**
This PR adds a script that automates the download of data on AP participation by state from the College Board's website. It was originally written to download data for the period 2007-2016 -- the second commit here has changes that allows someone to download data for a single year (in this case, 2017) and for multiple tests (in this case, CSP and CSA, since this is the first year that the two were offered).

It has no impact on any production data or code, but is a resource that will be used annually to update the figures here:
https://code.org/promote/ap

### CODE REVIEW

**Comment by caleybrock** [comment]
I think it's important to record scripts that have actually been run here, so I don't think you should change the code, but I do think you should add basically your PR description to a comment in the top of the file so it's clear what's this is used for when someone comes across it.

**Comment by bencodeorg** [comment]
I don't totally understand...the script has been run, it just downloads/processes/outputs the data locally (or wherever is specified in the `OUTPUT_FILE`)

File: tools/scripts/college_board_data.rb [hunk 1]
```diff
@@ -256,11 +279,17 @@ def get_value(spreadsheet, year, state, tab, column_row)
...
+def get_xlss(specific_year=nil)
   STATES_WITH_DC.each do |state|
     puts "  DOWNLOADING #{state}..."
     SPECS.each do |spec|
-      spec[:years].each do |year|
+      if specific_year.nil?
+        year_collection = spec[:years]
+      else
+        year_collection = [specific_year]
+      end
```
Comment (asherkach) [inline tools/scripts/college_board_data.rb]: nit: Consider a ternary, e.g., `year_collection = specific_year.nil? ? spec[:years] : [specific_year]`.
Also below.

File: tools/scripts/college_board_data.rb [hunk 2]
```diff
@@ -256,11 +279,17 @@ def get_value(spreadsheet, year, state, tab, column_row)
 end
 
 # Downloads the XLS files from the AP College Board.
-def get_xlss
+# Assumes you want to extract data for all years unless a year is specified
+def get_xlss(specific_year=nil)
```
Comment (asherkach) [inline tools/scripts/college_board_data.rb]: nit: Update the documentation to reflect the new parameter.
```
# @param [Integer] specific_year The specific year, if any, to download. If nil, all years are downloaded.
#   Default is nil.
```
Also below.

File: tools/scripts/college_board_data.rb [hunk 3]
```diff
@@ -273,26 +302,39 @@ def get_xlss
...
-        spec[:cells].each do |tab, column_rows|
-          column_rows.each do |column_row|
-            data << [
-              state,
-              year,
-              tab,
-              column_row,
-              get_value(spreadsheet, year, state, tab, column_row)
-            ]
+        spec[:years].each do |i|
```
Comment (asherkach) [inline tools/scripts/college_board_data.rb]: nit: Please use something more descriptive than `i`. Perhaps `spec_year`.

**Comment by bencodeorg** [comment]
@asherkach fixed nits. @caleybrock overheard some discussion of tracking the version of scripts like these that was actually run, versus the one that reflects style changes like the ones I just made  -- is there any way for me to mark that PR (that was actually run)?

**Comment by caleybrock** [comment]
No I think what you did is good because you have versions of it in your various commits. 

I was asking about how to review code like this that has already been used, because to me we should be keeping exactly what was run for history. So I was hesitant to make code comments about changing the script at all. Andrew made a good point that since you have various versions in your commits, and the whole point of putting it in our repro is to use it again, it's fine to review/make improvements.

**Comment by bencodeorg** [comment]
Ok cool -- is there any way to add tags to commits (or edit the commit message, but that seems hard?)

**Pull request merged**